package com.blue.bank.blue_bank.infrastructure.config;
import com.blue.bank.blue_bank.domain.exception.AccountNotActiveException;
import com.blue.bank.blue_bank.domain.exception.AccountNotFoundExcepcion;
import com.blue.bank.blue_bank.domain.exception.InsufficientFundsException;
import com.blue.bank.blue_bank.domain.exception.FraudCheckException;
import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ProblemDetail;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import java.util.ArrayList;
import java.util.List;
@RestControllerAdvice
public class GlobalExceptionHandler {
    @ExceptionHandler(AccountNotFoundExcepcion.class)
    public ProblemDetail handleAccountNotFound(AccountNotFoundExcepcion ex){
        ProblemDetail problem = ProblemDetail.forStatusAndDetail(
                HttpStatus.NOT_FOUND, ex.getMessage()
        );
        problem.setTitle("Cuenta no encontrada");
        return problem;
    }
    @ExceptionHandler(InsufficientFundsException.class)
    public ProblemDetail handleInsufficienFundsException(InsufficientFundsException ex){
        ProblemDetail problem = ProblemDetail.forStatusAndDetail(
                HttpStatusCode.valueOf(422), ex.getMessage()
        );
        problem.setTitle("Fondos insuficientes");
        return problem;
    }
    @ExceptionHandler(AccountNotActiveException.class)
    public ProblemDetail handleAccountNotActiveException(AccountNotActiveException ex){
        ProblemDetail problem = ProblemDetail.forStatusAndDetail(
                HttpStatusCode.valueOf(422), ex.getMessage()
        );
        problem.setTitle("Cuenta no activa");
        return problem;
    }
    @ExceptionHandler(Exception.class)
    public ProblemDetail hadleGeneral(Exception ex){
        ProblemDetail problem = ProblemDetail.forStatusAndDetail(
                HttpStatus.INTERNAL_SERVER_ERROR, "Error interno del servidor"
        );
        problem.setTitle("Error interno");
        return problem;
    }
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ProblemDetail handleValidation(MethodArgumentNotValidException ex) {
        ProblemDetail problem = ProblemDetail.forStatus(HttpStatus.BAD_REQUEST);
        problem.setTitle("Error de validación");
        List<String> errors = new ArrayList<>();
        ex.getBindingResult().getFieldErrors()
                .forEach(error ->
                        errors.add(error.getField() + ": " + error.getDefaultMessage()));
        ex.getBindingResult().getGlobalErrors()
                .forEach(error ->
                        errors.add(error.getDefaultMessage()));
        problem.setProperty("errors", errors);
        return problem;
    }
    @ExceptionHandler(FraudCheckException.class)
    public ProblemDetail handleFraudCheck(FraudCheckException ex){
        ProblemDetail problem = ProblemDetail.forStatusAndDetail(
                HttpStatusCode.valueOf(422), ex.getMessage()
        );
        problem.setTitle("Operación bloqueada por fraude");
        return problem;
    }
}